File: logger/src/main/java/fr/beapp/logger/appender/FileAppender.java
package fr.beapp.logger.appender;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.PrintStream;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import fr.beapp.logger.Logger;

/**
 * Write log messages on Android's filesystem.
 * this class handle file rotation based on current day.
 */
public class FileAppender extends Appender {

	private final SimpleDateFormat dateTimeFormatter = new SimpleDateFormat("HH:mm:ssZZ", Locale.ENGLISH);
	private final PrintStream printStream;
	private final File outputFile;

	/**
	 * Possible placeholder for filenamePattern :
	 * <ul>
	 * <li><code>{date}</code> : replaced by current date</li>
	 * </ul>
	 */
	public FileAppender(@NonNull File logsDirectory, @NonNull String filenamePattern) throws FileNotFoundException {
		this(Log.INFO, logsDirectory, filenamePattern);
	}

	/**
	 * Possible placeholder for filenamePattern :
	 * <ul>
	 * <li><code>{date}</code> : replaced by current date</li>
	 * </ul>
	 */
	public FileAppender(@Logger.LogLevel int level, @NonNull File logsDirectory, @NonNull String filenamePattern) throws FileNotFoundException {
		super(level);
		outputFile = ensureOutputFile(logsDirectory, buildFilename(filenamePattern));
		printStream = ensurePrintStream(outputFile);
	}

	/**
	 * Close the stream opened to write the file
	 */
	public void close() {
		printStream.close();
	}

	@Override
	public void log(@Logger.LogLevel int priority, @NonNull String message, @Nullable Throwable tr) {
		try {
			printStream.print(buildLogline(priority, message));
			printStream.print('\n');
			printStream.flush();
		} catch (Exception ignored) {
			// Can't use Logger.error here
		}
	}

	@NonNull
	public File getOutputFile() {
		return outputFile;
	}

	@NonNull
	protected PrintStream ensurePrintStream(@NonNull File outputFile) throws FileNotFoundException {
		FileOutputStream fileOutputStream = new FileOutputStream(outputFile, false);
		return new PrintStream(fileOutputStream, true);
	}

	@NonNull
	protected File ensureOutputFile(@NonNull File logsDirectory, @NonNull String filename) throws FileNotFoundException {
		File logFile = new File(logsDirectory, filename);
		File logFolder = logFile.getParentFile();

		if (logFolder.exists() || logFolder.mkdirs()) {
			return logFile;
		}
		throw new FileNotFoundException("Couldn't ensure log file '" + logFile + "'");
	}

	@NonNull
	protected String buildFilename(@NonNull String filenamePattern) {
		Date now = new Date();
		String date = new SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH).format(now);
		String time = new SimpleDateFormat("HH-mm-ss", Locale.ENGLISH).format(now);

		String filename = filenamePattern;
		filename = filename.replace("{date}", date);
		filename = filename.replace("{time}", time);
		return filename;
	}

	@NonNull
	protected String buildLogline(@Logger.LogLevel int priority, @NonNull String message) {
		return String.format("%s %s[%s]: %s", dateTimeFormatter.format(new Date()), Logger.findLevelName(priority), Thread.currentThread().getName(), message);
	}

}

File: logger/src/main/java/fr/beapp/logger/appender/FileAppender.kt
package fr.beapp.logger.appender

import android.util.Log
import fr.beapp.logger.Logger
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.PrintStream
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Write log messages on Android's filesystem.
 * this class handle file rotation based on current day.
 */
public open class FileAppender
/**
 * Possible placeholder for filenamePattern :
 * <ul>
 * <li><code>{date}</code> : replaced by current date</li>
 * </ul>
 */
@Throws(FileNotFoundException::class)
@JvmOverloads
constructor(
	@Logger.LogLevel logLevel: Int = Log.INFO,
	logsDirectory: File,
	filenamePattern: String
) : Appender(logLevel) {

	private val dateTimeFormatter: SimpleDateFormat = SimpleDateFormat("HH:mm:ssZZ", Locale.ENGLISH)
	private val outputFile: File = ensureOutputFile(logsDirectory, buildFilename(filenamePattern))
	private val printStream: PrintStream = ensurePrintStream(outputFile)

	/**
	 * Close the stream opened to write the file
	 */
	public fun close() {
		printStream.close()
	}

	override fun log(priority: Int, message: String, tr: Throwable?) {
		try {
			printStream.print(buildLogline(priority, message))
			printStream.print('\n')
			printStream.flush()
		} catch (ignored: Exception) {
			// Can't use Logger.error here
		}
	}

	public fun getOutputFile(): File = outputFile

	@Throws(FileNotFoundException::class)
	protected open fun ensurePrintStream(outputFile: File): PrintStream {
		val fileOutputStream = FileOutputStream(outputFile, false)
		return PrintStream(fileOutputStream, true)
	}

	@Throws(FileNotFoundException::class)
	protected open fun ensureOutputFile(logsDirectory: File, filename: String): File {
		val logFile = File(logsDirectory, filename)
		val logFolder = logFile.parentFile ?: throw FileNotFoundException("Couldn't create log file '$filename'")

		if (logFolder.exists() || logFolder.mkdirs()) {
			return logFile
		}
		throw FileNotFoundException("Couldn't ensure log file '$logFile'")
	}

	protected open fun buildFilename(filenamePattern: String): String {
		val now = Date()
		val date = SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH).format(now)
		val time = SimpleDateFormat("HH-mm-ss", Locale.ENGLISH).format(now)

		return filenamePattern
			.replace("{date}", date)
			.replace("{time}", time)
	}

	protected open fun buildLogline(@Logger.LogLevel priority: Int, message: String): String {
		return String.format(
			"%s %s[%s]: %s", dateTimeFormatter.format(Date()),
			Logger.findLevelName(priority),
			Thread.currentThread().name,
			message
		)
	}

}